docs/compaction-merge-loops.md

```diff
@@ -0,0 +1,153 @@
+# Compaction Merge Loops
+
+Compaction operations transform a set of input kvsets into a set of
+new output kvsets.  They are used are used to manage CN Tree shape
+(fanout, node size, node length), and to control space amp (e.g., by
+eliminating deleted key-value pairs).
+
+There are three types of compaction operations:
+- spill: consume one or more kvsets from the root node, produce at most
+  one output kvset for each leaf node.
+- kv-compact: consume two or more kvsets, produce one kvset
+- k-compact: similar to kv-compact, but without compacting vblocks
+
+HSE employs two additional operations that are an integral part of CN
+Tree maintenance but strictly speaking are not compaction operations:
+- split: split one kvset into two kvsets
+- join: join two kvsets into one kvset
+
+While split and join do transform input kvsets into a set of new
+output kvsets, they are meta-data-only operations.  They don't rewrite
+kblocks or vblocks and they don't use compaction merge loops.  This
+document does not apply to split or join operations.
+
+Spill, k-compact and kv-compact operations follow these steps:
+```
+  1. read entries from one or more input kvsets
+  2. merge input entries into a single stream of entries ordered by key
+  3. read ordered stream of entries to produce a single output stream of entries
+  4. write output entries to one or more output kvsets
+  5. commit output kvsets, making them visible to queries
+  6. remove input kvsets from the query path
+  7. delete input kvsets
+
+Figure 1: Compaction Steps
+```
+
+This document focuses on step 3, which is what we all the "merge loop"
+(i.e., the loop that processes the merged input stream).
+
+## A Model
+
+Each value has a type and sequence number. There are three
+value types: regular (V), tombstone (T), and prefix tombstone (PT).
+Since compaction logic cares about value types and sequence numbers,
+but not actual values, the simplified notation `<type><n>` can be used
+to represent a value with type `<type>` and sequence number `<n>`.
+
+Terminology:
+- A `value-type` is one of `V`, `T` or `PT`.
+- A `seqno` is an integer greater than 0.
+- A `value` is a 2-tuple of `(value-type, seqno)`.
+- A `key` is a string with length > 0.
+- An `entry` is a `key` with one or more `values`.
+
+Rules:
+- A kvset contains 1 or more entries.
+- A kvset cannot have two `V` or `T` entries with the same key.
+- A kvset cannot have two `PT` entries with the same key.
+
+Observations:
+- A kvset can have a `V` entry and a `PT` entry with the same key.
+
+## Condensed input stream
+
+We use the notion of a "condensed input stream" to be able to describe merge
+loops at a higher level.
+
+To understand what this means and why it is useful, consider the
+following two kvsets:
+```
```

Comment:
Apply a text label: \`\`\`text

Reply:
Isn't that the default?  I'd rather leave it out bc it clutters readability in plain text format.

Reply:
It is, but markdown linters will always get you :)